File: lua/application-window.lua
local lgi = require("lgi")
local Gtk = lgi.require("Gtk", "3.0")

-- GtkApplicationWindow: A GtkWindow with GtkApplication integration

local app_id = "io.github.Miqueas.GTK-Examples.Lua.Gtk3.ApplicationWindow"
local app = Gtk.Application({ application_id = app_id })

function app:on_startup()
  -- Since we're not going to use this window, I don't store it in a variable
  Gtk.ApplicationWindow({
    -- By setting the application to the new window is enough to to present it later
    application = self,
    default_width = 400,
    default_height = 400
  })
end

function app:on_activate()
  -- Because we set the application to the window (line 13), now we have a "default/main"
  -- window to present
  self.active_window:present()
end

return app:run(arg)

File: cpp/gtk3/oop/application-window.cpp
#include <gtkmm.h>

static const Glib::ustring APP_ID = "io.github.Miqueas.GTK-Examples.Cpp.Gtk3.OOP.ApplicationWindow";
static const Glib::ustring APP_TITLE = "Gtk::ApplicationWindow";

class App : public Gtk::Application {
  public:
    static Glib::RefPtr<App> create() {
      return Glib::RefPtr<App>(new App());
    }

  protected:
    App() : Gtk::Application(APP_ID) {}

    void on_activate() override {
      Gtk::Window* window = get_active_window();
      if (window) window->present();
    }

    void on_startup() override {
      // Ensures chain-up
      Gtk::Application::on_startup();

      auto window = new Gtk::ApplicationWindow();
      window->set_title(APP_TITLE);
      window->set_default_size(400, 400);

      add_window(*window);
    }
};

int main(int argc, char** argv) {
  return App::create()->run(argc, argv);
}